### PR TITLE
En 3609 bugfix interceptors should not broadcast when a direct msg has been received

### PR DESCRIPTION
**Pull request by iulianpascalau**
changed MessageProcessor's signature
removed bug when a direct message received could cause broatcast on the entire network
added test taht shows this bug has been removed

### CODE REVIEW

File: p2p/libp2p/discovery/mdnsDiscoverer_test.go [hunk 1]
```diff
@@ -114,7 +114,7 @@ func TestMdnsPeerDiscoverer_ApplyContextShouldWork(t *testing.T) {
 	//TODO delete skip when mdns library is concurrent safe
 	t.Skip("mdns library is not concurrent safe (yet)")
 	ctx, _ := libp2p2.NewLibp2pContext(context.Background(), &mock.ConnectableHostStub{})
-	mdns := discovery.NewMdnsPeerDiscoverer(time.Duration(time.Second*1), "srvcTag")
+	mdns := discovery.NewMdnsPeerDiscoverer(time.Second*1, "srvcTag")
```
Comment (sasurobert) [inline p2p/libp2p/discovery/mdnsDiscoverer_test.go]: time.second is enough "*1" is in plus

File: p2p/libp2p/netMessenger_test.go [hunk 2]
```diff
@@ -76,6 +76,30 @@ func createMockNetworkOf2() (mocknet.Mocknet, p2p.Messenger, p2p.Messenger) {
 	return netw, mes1, mes2
 }
 
+func createMockNetworkOfN(numOfPeers int) (mocknet.Mocknet, []p2p.Messenger) {
```
Comment (sasurobert) [inline p2p/libp2p/netMessenger_test.go]: createMockNetwork

File: p2p/libp2p/netMessenger_test.go [hunk 3]
```diff
@@ -1187,6 +1211,65 @@ func TestLibp2pMessenger_SendDataThrottlerShouldReturnCorrectObject(t *testing.T
...
+	idxResolver := 1
+	fmt.Println("broadcasting a message")
+	peers[idxResolver].Broadcast(topic, broadcastMsgResolver)
+
+	time.Sleep(time.Second)
+
+	fmt.Println("sending a direct message")
+	_ = peers[idxResolver].SendToConnectedPeer(topic, directMsgResolver, peers[idxRequester].ID())
+
+	time.Sleep(time.Second * 2)
```
Comment (sasurobert) [inline p2p/libp2p/netMessenger_test.go]: add some final assert check, that the code was reached as many time as you hoped

File: p2p/memp2p/memp2p.go [hunk 4]
```diff
@@ -334,7 +334,7 @@ func (messenger *Messenger) SendToConnectedPeer(topic string, buff []byte, peerI
 // previously registered a message processor for that topic. The Network will
 // log the message only if the Network.LogMessages flag is set and only if the
 // Messenger has the requested topic and MessageProcessor.
-func (messenger *Messenger) ReceiveMessage(topic string, message p2p.MessageP2P) error {
+func (messenger *Messenger) ReceiveMessage(topic string, message p2p.MessageP2P, allowBroadcast bool) error {
```
Comment (sasurobert) [inline p2p/memp2p/memp2p.go]: broadcast is enough for parameter name.

File: p2p/memp2p/memp2p.go [hunk 5]
```diff
@@ -351,7 +351,14 @@ func (messenger *Messenger) ReceiveMessage(topic string, message p2p.MessageP2P)
...
+	var handler func(buffToSend []byte)
+	if allowBroadcast {
+		handler = func(buffToSend []byte) {
+			messenger.Broadcast(topic, buffToSend)
+		}
+	}
+
+	err := validator.ProcessReceivedMessage(message, handler)
 
 	return err
```
Comment (sasurobert) [inline p2p/memp2p/memp2p.go]: return directly. no need to save in error

File: consensus/spos/worker.go [hunk 6]
```diff
@@ -210,7 +210,7 @@ func (wrk *Worker) getCleanedList(cnsDataList []*consensus.Message) []*consensus
 }
 
 // ProcessReceivedMessage method redirects the received message to the channel which should handle it
-func (wrk *Worker) ProcessReceivedMessage(message p2p.MessageP2P) error {
+func (wrk *Worker) ProcessReceivedMessage(message p2p.MessageP2P, _ func(buffToSend []byte)) error {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply consensus/spos/worker.go]: _ has been used here so we do not get weak warning stating the parameter is unused.

File: p2p/libp2p/discovery/mdnsDiscoverer_test.go (hunk 1 above)
Comment (iulianpascalau) [reply p2p/libp2p/discovery/mdnsDiscoverer_test.go]: agree

Comment (iulianpascalau) [reply p2p/libp2p/discovery/mdnsDiscoverer_test.go]: Searched and removed from entire project

File: p2p/libp2p/netMessenger_test.go (hunk 2 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go]: renamed

File: p2p/libp2p/netMessenger_test.go (hunk 3 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go]: Good catch! It could otherwise give false positive results.

Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go]: done

File: p2p/memp2p/memp2p.go (hunk 4 above)
Comment (iulianpascalau) [reply p2p/memp2p/memp2p.go]: I would keep this allow part as to be more clear of what "broadcast" should mean.

File: p2p/memp2p/memp2p.go (hunk 5 above)
Comment (iulianpascalau) [reply p2p/memp2p/memp2p.go]: changed

**Pull request merged**